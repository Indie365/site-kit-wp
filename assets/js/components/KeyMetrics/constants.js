/**
 * Key Metrics components - constants.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const KEY_METRICS_SETUP_CTA_WIDGET_SLUG = 'key-metrics-setup-cta-widget';
export const KEY_METRICS_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-key-metrics-selection-panel-opened';
export const KEY_METRICS_SELECTION_FORM = 'key-metrics-selection-form';
export const KEY_METRICS_SELECTED = 'key-metrics-selected';
export const MIN_SELECTED_METRICS_COUNT = 2;
export const MAX_SELECTED_METRICS_COUNT = 4;

export const KEY_METRICS_GROUP_VISITORS = {
	SLUG: 'visitors',
	LABEL: 'Visitors',
};
export const KEY_METRICS_GROUP_DRIVING_TRAFFIC = {
	SLUG: 'drivingTraffic',
	LABEL: 'Driving traffic',
};
export const KEY_METRICS_GROUP_GENERATING_LEADS = {
	SLUG: 'generatingLeads',
	LABEL: 'Generating leads',
};
export const KEY_METRICS_GROUP_SELLING_PRODUCTS = {
	SLUG: 'sellingProducts',
	LABEL: 'Selling products',
};
export const KEY_METRICS_GROUP_CONTENT_PERFORMANCE = {
	SLUG: 'contentPerformance',
	LABEL: 'Content performance',
};
