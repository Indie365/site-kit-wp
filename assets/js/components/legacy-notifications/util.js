/**
 * Notification utility functions.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { applyFilters } from '@wordpress/hooks';

export const wincallbacks = applyFilters( 'googlesitekit.winCallbacks', {} );

/**
 * Retrieves the total number of notifications from session storage.
 *
 * Otherwise make the requests to get modules notifications.
 *
 * @since 1.0.0
 *
 * @return {number} Total number of notifications.
 */
export async function getTotalNotifications() {
	const { setup } = global._googlesitekitLegacyData;

	if (
		! setup.isSiteKitConnected ||
		( ! setup.isAuthenticated && ! setup.isVerified )
	) {
		return 0;
	}

	// Will always return 0 until a new notification system is in place.
	const total = 0;

	// The total notifications count should always rely on local storage
	// directly for external availability.
	if ( global.localStorage ) {
		global.localStorage.setItem(
			'googlesitekit::total-notifications',
			total
		);
	}

	return total;
}

export const incrementCount = ( state ) => {
	const value = Math.abs( state.count ) + 1;
	// The total notifications count should always rely on local storage
	// directly for external availability.
	if ( global.localStorage ) {
		global.localStorage.setItem(
			'googlesitekit::total-notifications',
			value
		);
	}
	return {
		count: value,
	};
};

export const decrementCount = ( state ) => {
	const value = Math.max( 0, Math.abs( state.count ) - 1 );
	// The total notifications count should always rely on local storage
	// directly for external availability.
	if ( global.localStorage ) {
		global.localStorage.setItem(
			'googlesitekit::total-notifications',
			value
		);
	}
	return {
		count: value,
	};
};
