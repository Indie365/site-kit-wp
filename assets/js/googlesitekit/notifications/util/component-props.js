/**
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import memize from 'memize';

/**
 * Internal dependencies
 */
import Notification from '../components/Notification';
import Dismiss from '../components/Notification/Dismiss';

/**
 * Gets the props to pass to a Notification's component.
 *
 * @since n.e.x.t
 *
 * @param {string} id The Notification's ID.
 * @return {Object} Props to pass to the Notification component.
 */
export const getNotificationComponentProps = memize( ( id ) => {
	return {
		Notification: withNotificationID( id )( Notification ),
		Dismiss: withNotificationID( id )( Dismiss ),
	};
} );

function withNotificationID( id ) {
	return ( WrappedComponent ) => {
		function WithNotificationID( props ) {
			return <WrappedComponent { ...props } id={ id } />;
		}
		WithNotificationID.displayName = 'WithNotificationID';
		if ( WrappedComponent.displayName || WrappedComponent.name ) {
			WithNotificationID.displayName += `(${
				WrappedComponent.displayName || WrappedComponent.name
			})`;
		}
		return WithNotificationID;
	};
}
