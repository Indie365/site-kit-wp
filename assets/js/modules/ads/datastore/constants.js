/**
 * `modules/ads` data store constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const MODULES_ADS = 'modules/ads';

export const ADWORDS_SCOPE = 'https://www.googleapis.com/auth/adwords';

export const PAX_SETUP_STEP = {
	LAUNCH: 1,
	FINISHED: 2,
};

// Date range offset days for Ads report requests.
export const DATE_RANGE_OFFSET = 1;
