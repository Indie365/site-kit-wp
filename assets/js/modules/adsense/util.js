/**
 * AdSense utility functions.
 *
 * Site Kit by Google, Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import data from 'GoogleComponents/data';
import {
	getSiteKitAdminURL,
	getReAuthUrl,
	sendAnalyticsTrackingEvent,
} from 'GoogleUtil';

const { each, find, filter } = lodash;
const { __, sprintf } = wp.i18n;

export function reduceAdSenseData( rows ) {
	const dataMap = [
		[
			{ type: 'date', label: 'Day' },
			{ type: 'number', label: 'RPM' },
			{ type: 'number', label: 'Earnings' },
			{ type: 'number', label: 'Impressions' },
		]
	];

	each( rows, ( row ) => {
		const date = new Date( row[0] );
		dataMap.push( [
			date,
			row[2],
			row[1],
			row[3],
		] );
	} );

	return {
		dataMap,
	};
}

/**
 * Determine the AdSense account status.
 *
 * @param {function} statusUpdateCallback The function to call back with status updates.
 */
export async function getAdSenseAccountStatus( statusUpdateCallback, existingTag = false ) {

	/**
	 * Defines the account status variables.
	 */
	let accountStatus      = '';
	let statusMessage      = '';
	let profile            = false;
	let ctaLink            = '';
	let ctaLinkText        = '';
	let ctaTarget          = false;
	let helpLink           = '';
	let helpLinkText       = '';
	let setupComplete      = false;
	let statusHeadline     = '';
	let issue              = '';
	let notice             = '';
	let icon               = '';
	let buttonLink         = false;
	let footerText         = '';
	let footerAppendedText = '';
	let footerCTA          = '';
	let footerCTALink      = '';
	let continueAction     = false;
	let accountTagMatch    = false;
	let clientId           = false;
	let switchLabel        = '';
	let tracking           = false;
	let switchOffMessage   = '';
	let switchOnMessage    = '';

	const { accountURL, signupURL } = googlesitekit.modules.adsense;

	try {

		// First, fetch the list of accounts connected to this user.
		statusUpdateCallback( __( 'Locating accounts…', 'google-site-kit' ) );
		const results     = await data.get( 'modules', 'adsense', 'accounts' ).then( res => res ).catch( e => e );
		const accountData = results.data && ( ! results.data.status || 200 === results.data.status  )  ? results.data : results;
		const hasError    = accountData && accountData.message && accountData.message.error;
		let id            = accountData && accountData.length && accountData[0] ? accountData[0].id : false;

		/**
		 * Handle error states.
		 */
		if ( ! accountData || ! id || hasError ) {

			const { errors } = accountData.message.error;
			const { reason } = errors[0];

			/**
			 * Status: noAdSenseAccount.
			 *
			 * No account.
			 */
			if ( 'noAdSenseAccount'  === reason || ! accountData || ! id ) {

				/**
				 * Status disapprovedAccount.
				 *
				 * There is an AdSense account, it is disapproved, suspended, terminated etc.
				 */
				if ( 'disapprovedAccount' === reason ) {
					accountStatus  = 'account-disapproved';
					statusHeadline =  __( 'Your site isn’t ready to show ads yet', 'google-site-kit' );
					statusMessage  =  __( 'You need to fix some things before we can connect Site Kit to your AdSense account.', 'google-site-kit' );
					ctaLinkText    = __( 'Go to AdSense to find out how to fix the issue', 'google-site-kit' );
					ctaLink        = accountURL;
				} else {

					// There is no AdSense account, there is an existing tag.
					if ( existingTag ) {
						accountStatus  = 'no-account-tag-found';
						statusHeadline =  __( 'Looks like you’re already using AdSense', 'google-site-kit' );
						statusMessage  =  __( 'We’ve found some AdSense code on your site, but it’s not linked to this Google account.', 'google-site-kit' );
						profile        = false;
						ctaLinkText    = __( 'Switch Google account', 'google-site-kit' );
						ctaLink        = getReAuthUrl( 'adsense', true );
						buttonLink     = true;
						switchLabel  = __( 'Let Site Kit place code on your site to get your site approved', 'google-site-kit' );
						continueAction = {
							statusHeadline: __( 'Create a new AdSense account', 'google-site-kit' ),
							statusMessage: __( 'Site Kit will place additional AdSense code on every page across your site after you create an account. This means Google will automatically place ads for you in all the best places.', 'google-site-kit' ),
							notice: __( 'We recommend you remove the old AdSense code from this site.', 'google-site-kit' ),
							icon: 'warning',
							continueText: __( 'Continue anyway', 'google-site-kit' ),
							ctaLinkText: __( 'Create AdSense Account', 'google-site-kit' ),
							ctaLink: signupURL,
							ctaTarget: '_blank',
							continueAction: false,
						};

					} else {
						accountStatus      = 'no-account';
						statusHeadline     =  __( 'Create your AdSense account', 'google-site-kit' );
						statusMessage      =  __( 'Site Kit will place AdSense code on every page across your site. This means Google will automatically place ads for you in all the best places.', 'google-site-kit' );
						profile            = true;
						ctaLinkText        = __( 'Create AdSense Account', 'google-site-kit' );
						ctaLink            = signupURL;
						ctaTarget          = '_blank';
						buttonLink         = true;
						footerText         = __( 'Already have an AdSense account?', 'google-site-kit'  );
						footerAppendedText = __( 'to connect to it', 'google-site-kit'  );
						footerCTA          = __( 'Switch Google account', 'google-site-kit' );
						footerCTALink      = getReAuthUrl( 'adsense', true );
						tracking           = {
							eventCategory: 'adsense_setup',
							eventName: 'create_adsense_account'
						};
					}
				}
			}
		} else {

			// Found one or more accounts for this user, continue processing.
			const accounts = accountData;

			// If multiple accounts are returned, we need to search thru all of them to find accounts with matching domains.
			if ( 1 < accounts.length ) {

				// Find accounts with a matching URL channel.
				statusUpdateCallback( __( 'Searching for domain…', 'google-site-kit' ) );
				for ( let account of accounts ) {
					const accountId = account.id;
					const urlchannels = await data.get( 'modules', 'adsense', 'urlchannels', { clientId: accountId } ).then( res => res ).catch( e => e );
					const parsedUrl = new URL( googlesitekit.admin.siteURL );
					const matches = urlchannels && urlchannels.length ? filter( urlchannels, { urlPattern: parsedUrl.hostname } ) : [];

					if ( 0 === matches.length ) {
						accountStatus  = 'account-pending-review';
						issue          = 'accountPendingReview';
						sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_pending', 'accountPendingReview status account-pending-review' );
					} else {
						id = matches[0].id;
						sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_detected' );
					}
				}
			}

			// Set AdSense account link with account found.
			googlesitekit.modules.adsense.accountURL = sprintf( 'https://www.google.com/adsense/new/%s/home', id );

			statusUpdateCallback( __( 'Account found, checking account status…', 'google-site-kit' ) );
			let results    = await data.get( 'modules', 'adsense', 'alerts', { accountId: id } ).then( res => res ).catch( e => e );
			const alerts   = results.data && ( ! results.data.status || 200 === results.data.status  ) ? results.data : results;
			const hasError = alerts && alerts.message && alerts.message.error;
			if ( find( results, { type: 'GRAYLISTED_PUBLISHER' } ) ) {
				accountStatus  = 'ads-display-pending';
				issue          = 'accountPendingReview';
				sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_pending', 'accountPendingReview status ads-display-pending' );
			} else {


				// Attempt to retrieve and save the client id.
				const clientResults  = await data.get( 'modules', 'adsense', 'clients' ).then( res => res ).catch( e => e );
				const clients = clientResults.data && ( ! clientResults.data.status || 200 === clientResults.data.status  ) ? clientResults.data : clientResults;
				const hasClientError = clients && clients.message && clients.message.error;
				const item = clients && clients.length ? find( clients, { productCode: 'AFC' } ) : false;
				if ( item ) {
					clientId = item.id;

					// Save the client ID immediately so we can verify the site by inserting the tag.
					await data.set( 'modules', 'adsense', 'client-id', { clientId } ).then( res => res ).catch( e => e );
				}

				if ( hasError ) {
					const { reason } = alerts.message.error.errors[0];

					/**
					 * Status: accountPendingReview
					 */
					if ( 'accountPendingReview' === reason ) {

						/**
						 * Account setup still needs completion.
						 *
						 * The 'ads-display-pending' state shows the AdSenseInProcessStatus component.
						 */
						accountStatus = 'ads-display-pending';
						issue = 'accountPendingReview';
						sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_pending', 'accountPendingReview status ads-display-pending' );
					}

				} else {

					statusUpdateCallback( __( 'Looking for AdSense client…', 'google-site-kit' ) );

					/**
					 * Status: Account created, but cannot get the ad code yet.
					 */
					if ( hasClientError ) {

						/**
						 * Account setup still needs completion.
						 *
						 * The 'account-required-action' state shows the AdSenseInProcessStatus component.
						 */
						accountStatus  = 'account-required-action';
						issue          = 'accountRequiredAction';
						sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_required_action', 'accountRequiredAction status' );
					} else {
						if ( item ) {
							clientId = item.id;

							// Check the URL channels.
							statusUpdateCallback( __( 'Looking for site domain…', 'google-site-kit' ) );

							let urlchannels = await data.get( 'modules', 'adsense', 'urlchannels', { clientId } ).then( res => res ).catch( e => e );

							// Find a URL channel with a matching domain
							const matches = urlchannels && urlchannels.length && filter( urlchannels, ( channel ) => {
								return 0 < googlesitekit.admin.siteURL.indexOf( channel.urlPattern );
							} );

							// No domains found in the account, it is newly set up and domain addition is pending.
							if ( 0 === urlchannels.length ) {
								accountStatus  = 'ads-display-pending';
								issue          = 'accountPendingReview';
								sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_pending', 'accountPendingReview status ads-display-pending' );
							} else {

								// No URL matching the site URL is found in the account, the account is still pending.
								if ( ( ! matches ) || 0 === matches.length ) {
									accountStatus  = 'account-pending-review';
									issue          = 'accountPendingReview';
									sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_pending', 'accountPendingReview status account-pending-review' );
								} else {
									const moduleURL = getSiteKitAdminURL(
										'googlesitekit-module-adsense',
										{}
									);

									if ( existingTag ) {

										// AdSense existing tag id matches detected client id.
										if ( clientId === existingTag ) {

											/**
											 * No error, matched domain, account is connected.
											 *
											 * Existing tag detected, matching client id.
											 */
											accountStatus    = 'account-connected';
											issue            = false;
											icon             = 'alert';
											statusHeadline   =  __( 'Site Kit will place AdSense code to your site', 'google-site-kit' );
											statusMessage    =  __( 'This means Google will automatically place ads for you in all the best places.', 'google-site-kit' );
											ctaLinkText      = __( 'Continue', 'google-site-kit' );
											ctaLink          = moduleURL;
											buttonLink       = true;
											accountTagMatch  = true;
											sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_connected', 'existing_matching_tag' );
											switchLabel      = __( 'Let Site Kit place code on your site', 'google-site-kit' ),
											switchOffMessage = __( 'If you don’t let Site Kit place the code you may not get the best ads experience. You can set this up later on the Site Kit settings page.', 'google-site-kit' );
											switchOnMessage  = __( 'If you’ve already set up ads on your site, it may change how they appear. You can customize this later in AdSense.', 'google-site-kit' );

										} else {

											/**
											 * No error, matched domain, account is connected.
											 *
											 * Existing tag detected, non-matching client id.
											 */
											accountStatus  = 'account-connected-nonmatching';
											issue          = false;
											icon           = false;
											statusHeadline =  __( 'Your site has code from another AdSense account', 'google-site-kit' );
											statusMessage  =  __( 'We’ve found some AdSense code on your site, but it’s not linked to this AdSense account.', 'google-site-kit' );
											profile        = false;
											ctaLinkText    = __( 'Switch Google account', 'google-site-kit' );
											ctaLink        = getReAuthUrl( 'adsense', true );
											buttonLink     = true;
											continueAction = {
												accountStatus: 'account-connected',
												continueText: __( 'Continue anyway', 'google-site-kit' ),
												statusHeadline: __( 'Site Kit will place AdSense code on your site', 'google-site-kit' ),
												statusMessage: __( 'To connect your site to your AdSense account, Site Kit will place AdSense code on your site. For a better ads experience, you should remove AdSense code that’s not linked to this AdSense account.', 'google-site-kit' ),
												profile: true,
												ctaLink: moduleURL,
												ctaLinkText: __( 'Continue', 'google-site-kit' ),
												continueAction: false,
												switchLabel: __( 'Let Site Kit place code on your site', 'google-site-kit' ),
												switchOffMessage: __( 'You can let Site Kit do this later.', 'google-site-kit' ),
											};
											sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_connected', 'existing_non_matching_tag' );

										}

									} else {

										/**
										 * No error, matched domain, account is connected.
										 */
										accountStatus    = 'account-connected';
										issue            = false;
										icon             = false;
										statusHeadline   =  __( 'Looks like you’re already using AdSense', 'google-site-kit' );
										statusMessage    =  __( 'Site Kit will place AdSense code on your site to connect your site to AdSense and help you get the most out of ads. This means Google will automatically place ads for you in all the best places.', 'google-site-kit' );
										ctaLinkText      = __( 'Continue', 'google-site-kit' );
										ctaLink          = moduleURL;
										buttonLink       = true;
										tracking         = {
											eventCategory: 'adsense_setup',
											eventName: 'complete_adsense_setup'
										};
										switchLabel      = __( 'Let Site Kit place code on your site to get your site approved', 'google-site-kit' );
										switchOffMessage = __( 'If you’ve already got some AdSense code on your site, we recommend you use Site Kit to place code to get the most out of AdSense.', 'google-site-kit' );

										// Send a callback to set the connection status.
										statusUpdateCallback( __( 'Connecting…', 'google-site-kit' ) );

										// Track this event.
										sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_connected' );

										// Save the publisher clientId: AdSense setup is complete!
										await data.set( 'modules', 'adsense', 'setup-complete', { clientId } ).then( res => res ).catch( e => e );
									}
								}
							}

						} else {

							/**
							 * No AFC matching client was found.
							 *
							 * There is an AdSense account, but the AFC account is disapproved.
							 */
							accountStatus  = 'account-disapproved';
							issue          = __( 'There is an AdSense account, but the AFC account is disapproved', 'google-site-kit' );
							icon           = 'error';
							statusHeadline =  __( 'Create Account', 'google-site-kit' );
							statusMessage  =  __( 'Create an AdMob account, then open AdSense and try to upgrade.', 'google-site-kit' );
							ctaLinkText    = __( 'Create an AdMob Account', 'google-site-kit' );
							ctaLink        = 'https://google.com/admob';
						}
					}
				}
			}
		}

		let accounts = [];

		if ( accountData && accountData.length ) {
			accounts = accountData;
		}

		// Save the account status.
		await data.set( 'modules', 'adsense', 'account-status', { accountStatus } ).then( res => res ).catch( e => e );

		return ( {
			isLoading: false,
			accountStatus,
			statusMessage,
			accounts,
			profile,
			ctaLink,
			ctaLinkText,
			ctaTarget,
			helpLink,
			helpLinkText,
			error: false,
			setupComplete,
			statusHeadline,
			issue,
			notice,
			icon,
			buttonLink,
			footerCTALink,
			footerCTA,
			footerText,
			footerAppendedText,
			continueAction,
			accountTagMatch,
			clientId,
			existingTag,
			switchLabel,
			tracking,
			switchOffMessage,
			switchOnMessage,
		} );
	} catch ( err ) {
		return ( {
			isLoading: false,
			error: err.code,
			message: err.message,
		} );
	}
}

/**
 * Check if adsense is connected from Analytics API.
 *
 * @returns {boolean}
 */
export const isAdsenseConnectedAnalytics = async() => {
	const { active } = googlesitekit.modules.adsense;
	let adsenseConnect = true;

	if ( active ) {
		await data.get( 'modules', 'analytics', 'adsense' ).then( ( res ) => {
			if ( res ) {
				adsenseConnect = true;
			}
		} ).catch( ( err ) => {

			if ( 400 === err.code && 'INVALID_ARGUMENT' === err.message ) {
				adsenseConnect = false;
			}
		} );
	}

	return new Promise( ( resolve ) => {
		resolve( adsenseConnect );
	} );
};

/**
 * Check for any value higher than 0 in values from AdSense data.
 *
 * @param {array} data Data returned from the AdSense.
 * @returns {boolean}
 */
export const isDataZeroAdSense = ( data ) => {
	let totals = [];
	if ( data.totals ) {
		totals = data.totals;
	}

	// Look for any value > 0.
	totals = totals.filter( ( total ) => {
		return 0 < total;
	} );
	return 0 === totals.length;
};
