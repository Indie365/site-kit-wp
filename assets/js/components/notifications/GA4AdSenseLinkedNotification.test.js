/**
 * GA4AdSenseLinkedNotification component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import GA4AdSenseLinkedNotification from './GA4AdSenseLinkedNotification';
import {
	render,
	createTestRegistry,
	provideModules,
	muteFetch,
} from '../../../../tests/js/test-utils';
import { CORE_USER } from '../../googlesitekit/datastore/user/constants';
import { MODULES_ANALYTICS_4 } from '../../modules/analytics-4/datastore/constants';
import {
	VIEW_CONTEXT_MAIN_DASHBOARD,
	VIEW_CONTEXT_SETTINGS,
} from '../../googlesitekit/constants';
import { withNotificationComponentProps } from '../../googlesitekit/notifications/util/component-props';

const GA4_ADSENSE_LINKED_NOTIFICATION =
	'top-earning-pages-success-notification';

// eslint-disable-next-line jest/no-disabled-tests
describe.skip( 'GA4AdSenseLinkedNotification', () => {
	let registry;

	const fetchDismissItem = new RegExp(
		'^/google-site-kit/v1/core/user/data/dismiss-item'
	);
	const analyticsReport = new RegExp(
		'^/google-site-kit/v1/modules/analytics-4/data/report'
	);

	const NotificationWithComponentProps = withNotificationComponentProps(
		GA4_ADSENSE_LINKED_NOTIFICATION
	)( GA4AdSenseLinkedNotification );

	beforeEach( () => {
		registry = createTestRegistry();
		provideModules( registry );
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
			adSenseLinked: true,
		} );
		registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );
	} );

	it( 'does not render if AdSense module is not active', async () => {
		muteFetch( analyticsReport );

		fetchMock.postOnce( fetchDismissItem, {
			body: JSON.stringify( [ GA4_ADSENSE_LINKED_NOTIFICATION ] ),
			status: 200,
		} );

		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
			{
				active: false,
				connected: false,
				slug: 'adsense',
			},
		] );

		const { container, waitForRegistry } = render(
			<NotificationWithComponentProps />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);
		await waitForRegistry();

		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'does not render if AdSense and Analytics are not linked', async () => {
		muteFetch( analyticsReport );

		fetchMock.postOnce( fetchDismissItem, {
			body: JSON.stringify( [ GA4_ADSENSE_LINKED_NOTIFICATION ] ),
			status: 200,
		} );

		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
			adSenseLinked: false,
		} );
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
			{
				active: true,
				connected: true,
				slug: 'adsense',
			},
		] );

		const { container, waitForRegistry } = render(
			<NotificationWithComponentProps />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);
		await waitForRegistry();

		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'does not render if report has data, and it is not already dismissed', async () => {
		muteFetch( analyticsReport );

		fetchMock.postOnce( fetchDismissItem, {
			body: JSON.stringify( [ GA4_ADSENSE_LINKED_NOTIFICATION ] ),
			status: 200,
		} );

		registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );

		const { container, waitForRegistry } = render(
			<NotificationWithComponentProps />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);
		await waitForRegistry();

		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'does not render if already dismissed', async () => {
		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedItems( [ GA4_ADSENSE_LINKED_NOTIFICATION ] );

		const { container, waitForRegistry } = render(
			<NotificationWithComponentProps />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);
		await waitForRegistry();

		expect( fetchMock ).not.toHaveFetched(
			'/google-site-kit/v1/modules/analytics-4/data'
		);

		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'does not render when not on the main or entity dashboard', async () => {
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
			{
				active: true,
				connected: true,
				slug: 'adsense',
			},
		] );

		fetchMock.getOnce( analyticsReport, {
			body: {
				rowCount: null,
			},
			status: 200,
		} );

		const { container, waitForRegistry } = render(
			<NotificationWithComponentProps />,
			{
				registry,
				viewContext: VIEW_CONTEXT_SETTINGS,
			}
		);
		await waitForRegistry();

		expect( fetchMock ).not.toHaveFetched(
			'/google-site-kit/v1/modules/analytics-4/data'
		);

		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'renders when both Analytics and AdSense modules are active and linked', async () => {
		fetchMock.getOnce( analyticsReport, {
			body: {
				rowCount: null,
			},
			status: 200,
		} );

		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
			{
				active: true,
				connected: true,
				slug: 'adsense',
			},
		] );

		registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );

		const { container, waitForRegistry } = render(
			<NotificationWithComponentProps />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);
		await waitForRegistry();

		expect( container ).toHaveTextContent(
			'Your AdSense and Analytics accounts are linked'
		);
		expect( container ).toMatchSnapshot();
	} );

	it( 'renders when report has no data and it was not previously dismissed', async () => {
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
			{
				active: true,
				connected: true,
				slug: 'adsense',
			},
		] );

		fetchMock.getOnce( analyticsReport, {
			body: {
				rowCount: null,
			},
			status: 200,
		} );

		const { container, waitForRegistry } = render(
			<NotificationWithComponentProps />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);
		await waitForRegistry();

		expect( container ).toHaveTextContent(
			'Your AdSense and Analytics accounts are linked'
		);
	} );
} );
