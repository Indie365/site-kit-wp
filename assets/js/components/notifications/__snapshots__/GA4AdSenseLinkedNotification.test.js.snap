// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GA4AdSenseLinkedNotification renders when both Analytics and AdSense modules are active and linked 1`] = `
<div>
  <div
    class="mdc-layout-grid"
  >
    <div
      class="mdc-layout-grid__inner"
    >
      <div
        class="googlesitekit-subtle-notification mdc-layout-grid__cell mdc-layout-grid__cell--align-middle mdc-layout-grid__cell--span-12"
      >
        <div
          class="googlesitekit-subtle-notification__icon"
        >
          <svg />
        </div>
        <div
          class="googlesitekit-subtle-notification__content"
        >
          <p>
            Your AdSense and Analytics accounts are linked.
          </p>
          <p
            class="googlesitekit-subtle-notification__secondary_description"
          >
            We’ll let you know as soon as there’s enough data available.
          </p>
        </div>
        <button
          class="mdc-button mdc-button--tertiary"
          target="_self"
        >
          <span
            class="mdc-button__label"
          >
            Got it
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
